lib/rmt/lockfile.rb

```diff
@@ -4,13 +4,13 @@ class RMT::Lockfile
 
   class << self
     def create_file
-      if File.exist?(LOCKFILE_LOCATION)
-        raise RMT::Lockfile::ExecutionLockedError
-      else
-        dirname = File.dirname(LOCKFILE_LOCATION)
-        FileUtils.mkdir_p(dirname) unless File.directory?(dirname)
-        File.open(LOCKFILE_LOCATION, 'w') { |f| f.write(Process.pid) }
-      end
+      dirname = File.dirname(LOCKFILE_LOCATION)
+      FileUtils.mkdir_p(dirname) unless File.directory?(dirname)
+
+      # has to be a class instance variable otherwise garbage-collected and lock is removed
+      @f = File.open(RMT::Lockfile::LOCKFILE_LOCATION, File::RDWR | File::CREAT)
+      @f.write(Process.pid.to_s) if @f.read.empty? # write PID if file is new. this has to be done before locking
```

Comment:
Here is some twisted logic I don't get:
Why check if the file was empty?
Afaik, you are using the flock mechanism to determine if another running instance has the lock on this file.

So why not just try to grab the lock, and if you get it, you write the pid into that file.
If the file is locked already, the exception is raised.

Because now, if you for some unexpected reason, already have a lock file with some random number in it, but no other flock on that file, when you start the mirror command, it will work and start mirroring.
If you start another instance of mirror, it will stop, because of the locked file, but it will tell you to kill the process with that random number, not the actual pid of the first mirror.

... I just tried it and ran into the same problem than what you commented on ("this has to be done before locking") ... Do you know why this is?

Is it maybe related to the not so nice work-around with the instance variable??
Dunno, but I don't like that construct too much anyway.

Why not do this more OO? You could have a RMT::Lockfile `instanciate` method which returns the file object. This object you can store in a variable (when calling it from the `locked_method` wrapper method). And when you give that class an instance method `lock!`, it would all read so much nicer imo: `RMT::Lockfile.new.lock!`

Of course these Lockfile instances also could have a `remove!` method.

Maybe by refactoring it like this, you can then write into that file while it is locked??
In it's current form, I couldn't figure out, why it is not working :thinking:

Reply:
I have a similar concern regarding the usage of the lockfile.
Right now there's this `locked_method` call in `lib/rmt/cli/base.rb`, which in turn calls `RMT::Lockfile.create_file` and `RMT::Lockfile.remove_file`.

Why not get rid of that, and instead move it INTO `RMT::Lockfile`?

Then in `lib/rmt/cli/main.rb` and `lib/rmt/cli/import.rb` you could instead do:

```ruby
  desc 'data PATH', 'Read SCC data from given path'
  def data(path)
    RMT::Lockfile.lock do # <===
      needs_path(path) do
        RMT::SCC.new(options).import(path)
      end
    end
  end
```

Then, within `RMT::Lockfile` you can also simply use `File.open do ...`, so the the file closes automatically at the end.

Reply:
@lagartoflojo I tried your suggestion, however maybe I understood it wrong - but it does not look any better IMHO. also another thing I changed - I unlock the file and write it to prevent he unexpected behavior described but @thutterer . so it looks like this:

ruby```
def lock(&_block)
      dirname = File.dirname(LOCKFILE_LOCATION)
      FileUtils.mkdir_p(dirname) unless File.directory?(dirname)
      # https://ruby-doc.org/core-2.5.0/File.html#method-i-flock
      File.open(RMT::Lockfile::LOCKFILE_LOCATION, File::RDWR | File::CREAT) do |f|
        if f.flock(File::LOCK_EX | File::LOCK_NB)
          f.flock(File::LOCK_UN) # unlock for writing
          f.write(Process.pid.to_s)
          f.flock(File::LOCK_EX) # lock again
        else
          pid = File.read(RMT::Lockfile::LOCKFILE_LOCATION)
          raise ExecutionLockedError.new(
            "Process is locked by the application with pid #{pid}. Close this application or wait for it to finish before trying again\n"
          )
        end

        begin
          yield
        ensure
          f.close
          File.delete(LOCKFILE_LOCATION) if File.exist?(LOCKFILE_LOCATION)
        end
      end
    end
```

Reply:
FYI: writing to file after locking it does not really writes to file however it even returns the number of bytes written. so that we can not know PID if we do so. I guess unlocking and locking it again fixes the problem